worker/src/handlers.rs

```diff
@@ -1,13 +1,163 @@
+    error::DagError,
+    metered_channel::{Receiver, Sender},
+    Batch, BatchDigest, PrimaryToWorker, PrimaryWorkerMessage, ReconfigureNotification,
+    WorkerBatchRequest, WorkerBatchResponse, WorkerMessage, WorkerPrimaryMessage,
+    WorkerSynchronizeMessage, WorkerToWorker,
 };
 
+#[cfg(test)]
+#[path = "tests/handlers_tests.rs"]
+pub mod handlers_tests;
+
+// Makes it possible via channels for anemo RPC handler functions to send child RPCs.
+// TODO: replace this once anemo adds support for access to the network within handlers.
+pub struct ChildRpcSender {
+    // The public key of this authority.
+    name: PublicKey,
+    // The id of this worker.
+    id: WorkerId,
+    // The worker information cache.
+    worker_cache: SharedWorkerCache,
+    // Timeout on RequestBatches RPC.
+    request_batches_timeout: Duration,
+    // Incoming RequestBatches requests to be sent out. Uses `unreliable_send` if a target is
+    // provided, or `lucky_broadcast` if a target node count is provided.
+    rx_request_batches_rpc: Receiver<(
+        Option<NetworkPublicKey>,
+        Option<usize>,
+        WorkerBatchRequest,
+        oneshot::Sender<Result<anemo::Response<WorkerBatchResponse>>>,
+    )>,
+    /// Receive reconfiguration updates.
+    rx_reconfigure: watch::Receiver<ReconfigureNotification>,
+    // Network to use for sending requests.
+    network: P2pNetwork,
+}
+
+impl ChildRpcSender {
+    #[must_use]
+    pub fn spawn(
+        name: PublicKey,
+        id: WorkerId,
+        worker_cache: SharedWorkerCache,
+        request_batches_timeout: Duration,
+        rx_request_batches_rpc: Receiver<(
+            Option<NetworkPublicKey>,
+            Option<usize>,
+            WorkerBatchRequest,
+            oneshot::Sender<Result<anemo::Response<WorkerBatchResponse>>>,
+        )>,
+        rx_reconfigure: watch::Receiver<ReconfigureNotification>,
+        network: P2pNetwork,
+    ) -> JoinHandle<()> {
+        tokio::spawn(async move {
+            Self {
+                name,
+                id,
+                worker_cache,
+                request_batches_timeout,
+                rx_request_batches_rpc,
+                rx_reconfigure,
+                network,
+            }
+            .run()
+            .await;
+        })
+    }
+
+    async fn handle_worker_batch_responses(
+        responses: Vec<Timeout<JoinHandle<Result<anemo::Response<WorkerBatchResponse>>>>>,
+        response_channel: oneshot::Sender<Result<anemo::Response<WorkerBatchResponse>>>,
+    ) {
+        let mut results: FuturesUnordered<_> = responses.into_iter().collect();
+        let mut last_result: Result<anemo::Response<WorkerBatchResponse>> = Err(anyhow::anyhow!(
+            "no WorkerBatchResponse received (or request(s) timed out)"
+        ));
+        while let Some(result) = results.next().await {
+            match result {
+                Ok(Ok(Ok(result))) => {
+                    let _ = response_channel.send(Ok(result));
```

Comment:
So here it seems we respond with the first result. But what happens to the other Tasks? I supose they keep running since they are not cancelled. Is that what we want or not? Since we have their join handle we can call https://docs.rs/tokio/latest/tokio/task/struct.JoinHandle.html#method.abort is this desirable?

Reply:
Yes good point. I avoided getting too fancy here since we want to delete `ChildRpcHandler` as soon as anemo supports child rpcs sent from handler functions